software/bootcamp-code/main.py

```diff
@@ -33,61 +108,257 @@ def main_loop():
+                if throw_duration>2 and ball_out_of_frame==True:
+                    state=State.FIND_BALL
+                    finder_timer=time.time()
+                    # reset variables
+                    throw_start=0
+                    ball_out_of_frame=False
+                    print("ball throwing time up")
+                    continue
+		
+                #when the ball is not in view, calculate proportional speed for the thrower and forward speed based on basket
+                if ball_out_of_frame==True:
+                    #print("ball out of frame, calculate side-speed prop to basket x location")
+                    x_speed_prop = norm_co(ball_desired_x, basket_to_throw.x, cam.rgb_width)
+                    rot_speed_prop = x_speed_prop
+
+                # when the ball is in view, drive towards it, x-speed based on ball and basket x-coordinate difference
+                else:
+                    #print("ball in view, calculate side-speed prop. to basket and ball x location difference")
+                    # x speed aka side speed is proportional to the distance of the ball from the basket
+                    x_speed_prop = norm_co(processedData.biggest_ball.x, basket_to_throw.x, cam.rgb_width)
+                    rot_speed_prop = norm_co(ball_desired_x, processedData.biggest_ball.x, cam.rgb_width)
+		    
+                # y speed aka forward speed is proportional to the basket distance in the frame considering y coordinate -destination is 100pixels from the bottom edge
+                y_speed_prop=norm_co((cam.rgb_height-100), basket_to_throw.y, (cam.rgb_height-100))
+                # calculate the thrower motor speed based on the basket distance gotten from depth camera
+                # take depth frame basket x, y depth, better to take matrix of all the nearest and the medium of that
+                basket_depth = get_depth(processedData.depth_frame, 0, basket_to_throw.x)
+                #print("BASKET_DEPTH:", basket_depth)
+                basket_dist_norm = basket_depth/max_basket_depth
+                thrower_speed_prop=basket_dist_norm*thrower_speed_range+throw_motor_speed_min
+		
+                omni_motion.move(-1*x_speed_prop*throw_move_speed, -1*y_speed_prop*throw_move_speed, rot_speed_prop*throw_move_speed*2, thrower_speed_prop) #added *2
+                
+            elif state==State.STOP:
+                omni_motion.move(0, 0, 0, 0)
+
+            # to test the thrower
+            elif state==State.TEST_THROWER:
+                omni_motion.move(0, 0, 0, testing_thrower_speed)
+                print("testing thrower")
+                processedData=processor.process_frame(aligned_depth=False)
+                basket_depth = get_depth(processedData.depth_frame, 0, basket_to_throw.x)
+                print(basket_depth)
+                #print(processor.process_frame(aligned_depth=False).basket_b.distance)
+
+            elif state==State.FIND_FURTHEST_BASKET:
```

Comment:
What is the point of this state? Could the drive to furthest basket not do the same since you already have a value in the basket_to_drive varible. You can just spin when not seeing the basket until you see it and then drive to it. It is quite hard not to see a basket. 

If you feel fancy you can also have a fallback case that when you dont see the basket either, it will swap baskets and if either basket is not visible, drive around and try not to drive out of the field.

Reply:
Well the basket color which to drive towards changes based on this state. It finds both baskets, compares their distance and drives towards the one that is further. The state stays as long as it takes to find both baskets and then to have the furthest basket in view. Ideally the robot makes half a spin, in other case it makes a full spin.
I guess I could hardcode it and say, that the furthest basket is always the one that is further than half a court away, (distance bigger than 2000).